rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,813 @@
+  </state.root>
+);
+```
+
+##### SlotComponent type & ComponentState change
+
+Some type changes will be required:
+
+###### Introduction of the `SlotComponent` type.
+
+Slot components are exotic components (like `React.forwardRef` and `React.memo` components),
+that will have intrinsically associated with them External properties, Internal properties and the base component type.
+
+```ts
+type SlotComponent<Props extends UnknownSlotProps = UnknownSlotProps> = React.ExoticComponent<
+  Props & React.RefAttributes<Element>
+> & {
+  readonly props: Props;
+  readonly defaultProps?: Partial<Props> | undefined;
+  readonly componentType:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+};
+```
+
+###### ComponentState
+
+`ComponentState` should map provided slots to `SlotComponent`s instead of mapping to resolved shorthands! And `components` property can be dropped.
+
+Before:
+
+```ts
+type ComponentState<Slots extends SlotPropsRecord> = {
+  components: {
+    [Key in keyof Slots]-?:
+      | React.ComponentType<ExtractSlotProps<Slots[Key]>>
+      | (ExtractSlotProps<Slots[Key]> extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  };
+} & {
+  // Include a prop for each slot, with the shorthand resolved to a props object
+  // The root slot can never be null, so also exclude null from it
+  [Key in keyof Slots]: ReplaceNullWithUndefined<
+    Exclude<Slots[Key], SlotShorthandValue | (Key extends 'root' ? null : never)>
+  >;
+};
+```
+
+After:
+
+```ts
+type ComponentState<Slots extends SlotPropsRecord> = {
+  [Key in keyof Slots]: SlotComponent<ExtractSlotProps<Slots[Key]>>;
+};
+```
+
+##### Styles hooks `.props` access
+
+A minor modification in style hooks are required. at the moment we mutate the merged properties provided by `resolveShorthand`
+to include our custom `className` property.
+
+In this case we have some alternatives:
+
+###### Option 1 mutate external properties
+
+Instead of mutating the merged properties, simply mutate the External properties from the slot component:
+
+```diff
+// useAccordionHeader
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
+  const styles = useStyles();
+- state.root.className = mergeClasses(
++ state.root.props.className = mergeClasses(
+    accordionHeaderClassNames.root,
+    styles.root,
+    state.inline && styles.rootInline,
+    state.disabled && styles.rootDisabled,
+-   state.root.className,
++   state.root.props.className,
+  );
+
+  if (state.icon) {
+-    state.icon.className = mergeClasses(
++    state.icon.props.className = mergeClasses(
+      accordionHeaderClassNames.icon,
+      styles.icon,
+-      state.icon.className
++      state.icon.props.className
+    );
+  }
+
+  return state;
+};
+```
+
+###### Option 2 mutate overrides
+
+By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated:
```

Comment:
We can always stick with option 1 and just keep mutating `props`. I'm particularly fond of option 3 though....

Reply:
Yes, I think mutating props is the best option IMO.
* Having separate `overrides` vs. `props` vs. `defaultProps` quickly gets difficult to manage.
* Having immutable `props` and copying the object on every override could have perf issues.

Reply:
Technically it would just be 2 objects on overrides, 1 for the styles and 1 for all overrides that come from the state hook, I don't see why more than that would be required, and they don't even need to be copied, we can merge them in the render method.